plugin/hover.py

```diff
@@ -209,15 +232,15 @@ def _show_hover(self, listener: AbstractViewListener, point: int, only_diagnosti
         _test_contents.append(contents)  # for testing only
 
         if contents:
+            # The previous popup could be in a different location from the next one
             if self.view.is_popup_visible():
-                update_lsp_popup(self.view, contents)
-            else:
-                show_lsp_popup(
-                    self.view,
-                    contents,
-                    flags=sublime.HIDE_ON_MOUSE_MOVE_AWAY,
-                    location=point,
-                    on_navigate=lambda href: self._on_navigate(href, point))
+                hide_lsp_popup(self.view)
```

Comment:
Shouldn't do that as that can result in popup position jumping while the content is arriving late.

I don't think you need this with the current implementation anyway?

Reply:
We need this because the hover and the actual selection could at different places see: ![Peek 2021-11-18 14-42](https://user-images.githubusercontent.com/1632384/142426666-5aa5a8a2-1c97-4a19-9d32-4de43738057c.gif)

Reply:
How is that with this selection:

```
|toFloat.toString()|
```

You are getting different results when hovering `toFloat` and `toString()`. Wouldn't the implementation send the same range in both cases?

Reply:
Looking close, it seems like your implement almost never sends range actually since you only take a range if "point" is not provided in LspHoverCommand but it almost always is.

So once you fix that we can investigate more but I'm still claiming that you shouldn't be hiding the popup as that will also affect the normal behavior negatively.

Reply:
Since there isn't an LSP spec the user behavior isn't defined. 
The current implementation sends only a hover point even when a selection is hovered. If you want to send the range you have to use the key binding. 
I guess you want the range to be sent if the user hovers a selection, right ?

@ckipp01 @tgodzik what is the user behavior in vscode and vim ?

Reply:
> I guess you want the range to be sent if the user hovers a selection, right ?

That's right, currently VS Code plugin always sends a range even with no selection, but this should work the same if point is sent.

Reply:
Yeah, I thought that's the point of the future.

And IMO this should work as expected first and foremost for the mouse use case. That is, if selection is hovered it should send a range to the server. If there is no selection under the hover point then I guess it doesn't really matter what is sent (point or range), the result should be the same.

Reply:
So why would it be necessary to hide the popup manually now? ST should be hiding it iself when hovering new point.

Reply:
Because then you get the pop at the wrong place.
1. Hover "App"
2. invoke `lsp_hover` for selection
3. pop up showing up in the wrong place

![Peek 2021-11-18 16-10](https://user-images.githubusercontent.com/1632384/142441954-409522b5-611f-4240-b2a3-e12e76fc2f4f.gif)

Reply:
It is an issue when using both. Generate a popup via mouse, keep the cursor in position and then invoke a hover via keyboard which leads to a popup showing in the wrong place. Hiding the popup fixes the problem because then we show the popup in the correct position

Reply:
OK so `mdpopups.update_popup` https://facelessuser.github.io/sublime-markdown-popups/api/#update-popup doesn't take location as argument so we have to hide it in this scenario

Reply:
So yes, it's an unrelated issue.
It should be fixed separately and in a different way.

Reply:
It is even a sublime API issue :(  https://www.sublimetext.com/docs/api_reference.html#ver-dev `update_popup` takes only `content` as arugment

Reply:
> what is the user behavior in vscode and vim ?

In `nvim-metals` if the user is in visual mode, it sends the range, if they are in normal mode it sends the point.

Reply:
That's fine because you can close the popup before showing it. The issue is not with closing the popup manually but with **when** it's closed. We don't want to close it when we've shown the popup with content from one session and we are waiting for the rest of the sessions. So this needs to be solved differently. Possibly by just closing the popup at the beginning of `LspHoverCommand`.

Reply:
But I don't want this unrelated fix to be mixed into this PR.

Reply:
ok will remove `hide_lsp_popup` and will create a separate issue once this is merged

Reply:
> 1. Hover "App" with mouse, popup shows up, keep cursor in place.
> 2. invoke `lsp_hover` for selection via key binding
> 3. Notice pop up showing up in the wrong place. The popup is still on "App" while it should move to "|toFloat.toString()|"

Shouldn't the popup for `lsp_hover` always be at the place where the mouse cursor is? So I would say when doing these steps, `lsp_hover` should ignore the selected range and still send the point from "App", regardless whether triggered by keybinding or not. Otherwise it wouldn't be "hover" anymore. Instead there should better be a new command for it (and ideally it shouldn't be handled by the `textDocument/hover` request if it's unrelated to mouse hover).

Reply:
Maybe I'm wrong because I interpreted the "hover" to always be related to the mouse hovering over a position, but what actually might be meant is that the popup is displayed hovering over the text ;)

Reply:
Please see https://github.com/microsoft/language-server-protocol/issues/377 the spec will most likely use `textDocument/hover`

Reply:
But from that discussion it seems my assumption was correct:

> What does the client send as position when a range is selected but cursor is not on the range? Just a Position, I presume?

> The current implementation just sends the position, yes.

So when doing the mentioned steps, `lsp_hover` should ignore the range and only send the position. And we don't need to care about updating popup locations.